addons/account/wizard/account_merge_wizard.py

```diff
@@ -0,0 +1,310 @@
+
+    @api.model
+    def default_get(self, fields):
+        res = super().default_get(fields)
+        if not set(fields) & {'account_ids', 'wizard_line_ids'} or set(res.keys()) & {'account_ids', 'wizard_line_ids'}:
+            return res
+
+        if self.env.context.get('active_model') != 'account.account':
+            raise UserError(_("This can only be used on accounts."))
+        if len(self.env.context.get('active_ids') or []) < 2:
+            raise UserError(_("You must select at least 2 accounts."))
+
+        res['account_ids'] = [Command.set(self.env.context.get('active_ids'))]
+        return res
+
+    def _get_grouping_key(self, account):
+        """ Return a grouping key for the given account. """
+        self.ensure_one()
+        grouping_fields = ['account_type', 'non_trade', 'currency_id', 'reconcile', 'deprecated']
+        if self.is_group_by_name:
+            grouping_fields.append('name')
+        return tuple(account[field] for field in grouping_fields)
+
+    def _get_group_name(self, account):
+        """ Return a human-readable name for an account's group, in the format:
+        '{Trade/Non-trade} Receivable {USD} {Reconcilable} {Deprecated}'
+        """
+        self.ensure_one()
+
+        account_type_label = dict(self.pool['account.account'].account_type._description_selection(self.env))[account.account_type]
+        if account.account_type in ['asset_receivable', 'liability_payable']:
+            account_type_label = _("Non-trade %s", account_type_label) if account.non_trade else _("Trade %s", account_type_label)
+
+        other_name_elements = []
+        if account.currency_id:
+            other_name_elements.append(account.currency_id.name)
+
+        if account.reconcile:
+            other_name_elements.append(_("Reconcilable"))
+
+        if account.deprecated:
+            other_name_elements.append(_("Deprecated"))
+
+        if not self.is_group_by_name:
+            grouping_key_name = account_type_label
+            if other_name_elements:
+                grouping_key_name = f'{grouping_key_name} ({", ".join(other_name_elements)})'
+        else:
+            grouping_key_name = f'{account.name} ({", ".join([account_type_label] + other_name_elements)})'
+
+        return grouping_key_name
+
+    @api.depends('is_group_by_name', 'account_ids')
+    def _compute_wizard_line_ids(self):
+        """ Determine which accounts to merge together. """
+        for (i, wizard) in enumerate(self):
+            # Filter out Bank / Cash accounts
+            accounts = wizard.account_ids._origin.filtered(lambda a: a.account_type not in ('asset_bank', 'asset_cash'))
+
+            wizard_lines_vals_list = []
+            sequence = 0
+            for grouping_key, group_accounts in accounts.grouped(wizard._get_grouping_key).items():
+                grouping_key_str = str((i, *grouping_key))  # The grouping key identifies both the wizard and the account properties
```

Comment:
You shouldn't include an id here, but use the wizard's id in the grouping key instead. (if even needed because you should never compare grouping keys from different wizards)

This `i` is stored, but depends on when (and with what) it was computed, which it cannot be.

Reply:
I agree this is useless, I only did it this way due to the constraints of the previous version where the wizard wasn't created before opening, so it only existed as a temporary record in an onchange, which means the id wasn't a reliable identifier. Removing it now.